File: 2016_MaximumDifferenceBetweenIncreasingElements/2016.c
int maximumDifference(int* nums, int numsSize) { 
    int maxDif = -1;

    for(int i = 0; i < numsSize-1; i++){
        for(int j = i+1; j < numsSize; j++){
            if(nums[i] > nums[j]) continue;

            int curDif = nums[j] - nums[i];

            if(curDif > maxDif && curDif != 0) maxDif = curDif;
        }
    }

    return maxDif;
}

File: 2016_MaximumDifferenceBetweenIncreasingElements/2016.exs
defmodule Solution do
  @spec maximum_difference(nums :: [integer]) :: integer
  def maximum_difference(nums) do
    [min | tail] = nums 

    solver(min, tail, -1)
  end

  # haskell like pattern matching SO KOOL
  def solver(min, [], maxDiff), do: maxDiff
  def solver(min, nums, maxDiff) do 
    [head | tail] = nums 

    if min >= head do 
      solver(head, tail, maxDiff)
    else 
      curDiff = head - min 
      solver(min, tail, max(maxDiff, curDiff))
    end
  end 
end